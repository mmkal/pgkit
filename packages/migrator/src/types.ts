import {sql, Client, Queryable} from '@pgkit/client'
import {type MigraOptions} from '@pgkit/migra'
import * as umzug from 'umzug'

export interface MigratorContext {
  connection: Queryable
  sql: typeof sql
}

export type Awaitable<T> = T | Promise<T>

export type Migration = (params: umzug.MigrationParams<MigratorContext>) => Promise<void>

export type Confirm = (sql: string, options?: {readonly?: boolean}) => Awaitable<string | null>

export interface BaseListedMigration {
  name: string
  path: string
  content: string
  note?: string
}

export interface PendingMigration extends BaseListedMigration {
  status: 'pending'
}

export interface ExecutedMigration extends BaseListedMigration {
  status: 'executed'
  drifted: boolean
}

export type ListedMigration = PendingMigration | ExecutedMigration

export type RunnableMigration = umzug.RunnableMigration<MigratorContext>

export type Logger = {
  info: (...args: unknown[]) => void
  warn: (...args: unknown[]) => void
  error: (...args: unknown[]) => void
}

export type Task = <T>(name: string, fn: () => Promise<T>) => Promise<{result: T}>

export interface MigratorConfig {
  /** @pgkit/client instance */
  client: Client
  migrationsPath: string
  migrationTableName?: string | string[]
  /**
   * Whether to use `client.transaction(tx => ...)` or `client.connect(cn => ...)` when running up/down migrations
   * @default `transaction`
   */
  connectMethod?: 'transaction' | 'connect'

  /**
   * A function which wraps its callback, can be used to add logging before/after completion. Compatible with [tasuku](https://npmjs.com/package/tasuku), for example.
   * @example Using tasuku
   * ```ts
   * import task from 'tasuku'
   *
   * const migrator = new Migrator(___)
   *
   * migrator.useConfig({task}, async () => {
   *  await migrator.up()
   * })
   * ```
   *
   * @example Using a custom task function
   * ```ts
   * const migrator = new Migrator(___)
   *
   * migrator.useConfig({
   *    task: async (name, fn) => {
   *      migrator.logger.info('Starting', name)
   *      const result = await fn()
   *      migrator.logger.info('Finished', name)
   *      return {result}
   *    },
   *    async () => {
   *      await migrator.up()
   *    })
   * })
   * ```
   */
  task: Task
  logger: Logger

  /** flags passed to [migra](https://npmjs.com/package/@pgkit/migra) - which is called for various introspection tasks */
  defaultMigraOptions?: MigraOptions
}

export interface MigratorConstructorParams extends Omit<MigratorConfig, 'client' | 'task' | 'logger'> {
  client: string | Client
  task?: Task
  logger?: Logger
}
