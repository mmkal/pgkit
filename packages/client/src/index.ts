// codegen:start {preset: barrel}
export * from './client'
export * from './errors'
export * from './naming'
export * from './sql'
export * from './storage'
export * from './type-parsers'
export * from './types'
// codegen:end
