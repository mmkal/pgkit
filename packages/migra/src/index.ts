export {Changes} from './changes'
export {run, type MigraOptions} from './command'
export {Migration} from './migra'
export {Statements, UnsafeMigrationException} from './statements'

export {PostgreSQL} from '@pgkit/schemainspect'
