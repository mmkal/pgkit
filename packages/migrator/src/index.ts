import {createHash} from 'crypto'
import {readFileSync, writeFileSync, mkdirSync, readdirSync} from 'fs'
import {once, memoize} from 'lodash'
import {map, pick} from 'lodash/fp'
import {basename, dirname, join, extname} from 'path'
import * as Umzug from 'umzug'
import {sql, DatabasePoolType} from 'slonik'
import {raw} from 'slonik-sql-tag-raw'
import {inspect} from 'util'
import * as dedent from 'dedent'
import {EOL} from 'os'

export const supportedExtensions = ['sql', 'js', 'ts'] as const

export type SupportedExtension = typeof supportedExtensions[number]

export interface SlonikMigratorOptions {
  slonik: DatabasePoolType
  migrationsPath: string
  migrationTableName?: string
  log?: typeof console.log
  args?: string[]
  mainModule?: NodeModule
}

export interface MigrationParams {
  path: string
  slonik: DatabasePoolType
  sql: typeof sql
}

export type Migration = (params: MigrationParams) => PromiseLike<unknown>

export interface MigrationResult {
  file: string
  path: string
}

export interface SlonikMigratorCLI {
  up(migration?: string): Promise<MigrationResult[]>
  down(migration?: string): Promise<MigrationResult[]>
  create(migration: string): string
}

/** @private not meant for general use, since this is coupled to @see umzug */
type GetUmzugResolver = (
  params: MigrationParams,
) => {
  up: () => PromiseLike<unknown>
  down?: () => PromiseLike<unknown>
}

const scriptResolver: GetUmzugResolver = params => {
  const exportedMigrations: {up: Migration; down?: Migration} = require(params.path)
  return {
    up: () => exportedMigrations.up(params),
    down: exportedMigrations.down && (() => exportedMigrations.down!(params)),
  }
}

const sqlResolver: GetUmzugResolver = ({path, slonik, sql}) => ({
  up: () => slonik.query(sql`${raw(readFileSync(path, 'utf8'))}`),
  down: async () => {
    const downPath = join(dirname(path), 'down', basename(path))
    await slonik.query(sql`${raw(readFileSync(downPath, 'utf8'))}`)
  },
})

export const defaultResolvers = {
  sql: sqlResolver,
  js: scriptResolver,
  ts: scriptResolver,
}

export const setupSlonikMigrator = ({
  slonik,
  migrationsPath,
  migrationTableName = 'migration',
  log = memoize(console.log, JSON.stringify),
  mainModule,
}: SlonikMigratorOptions) => {
  const migrationResolver: GetUmzugResolver = params => {
    const ext = params.path.split('.').slice(-1)[0] as SupportedExtension
    return defaultResolvers[ext](params)
  }
  const createMigrationTable = once(async () => {
    void (await slonik.query(sql`
      create table if not exists ${sql.identifier([migrationTableName])}(
        name text primary key,
        hash text not null,
        date timestamptz not null default now()
      )
    `))
  })
  const hash = (migrationName: string) =>
    createHash('md5')
      .update(readFileSync(join(migrationsPath, migrationName), 'utf8').trim().replace(/\s+/g, ' '))
      .digest('hex')
      .slice(0, 10)

  const umzug = new Umzug({
    logging: log,
    migrations: {
      path: migrationsPath,
      pattern: /\.(sql|js|ts)$/,
      customResolver: path => migrationResolver({path, slonik, sql}),
    },
    storage: {
      async executed() {
        await createMigrationTable()
        return slonik
          .any(sql`select name, hash from ${sql.identifier([migrationTableName])}`)
          .then(migrations => {
            log('migrations in database:', migrations)
            return migrations
          })
          .then(migrations =>
            migrations.map(r => {
              const name = r.name as string
              /* istanbul ignore if */
              if (r.hash !== hash(name)) {
                log(
                  `warning:`,
                  `hash in '${migrationTableName}' table didn't match content on disk.`,
                  `did you try to change a migration file after it had been run?`,
                  {migration: r.name, dbHash: r.hash, diskHash: hash(name)},
                )
              }
              return name
            }),
          )
      },
      async logMigration(name: string) {
        await createMigrationTable()
        await slonik.query(sql`
            insert into ${sql.identifier([migrationTableName])}(name, hash)
            values (${name}, ${hash(name)})
          `)
      },
      async unlogMigration(name: string) {
        await createMigrationTable()
        await slonik.query(sql`
            delete from ${sql.identifier([migrationTableName])}
            where name = ${name}
          `)
      },
    },
  })

  const templates = (name: string) => ({
    sql: {
      up: `--${name} (up)`,
      down: `--${name} (down)`,
    },
    ts: {
      up: dedent`
        import {Migration} from '@slonik/migrator'

        export const up: Migration = ({slonik, sql}) => slonik.query(sql\`select true\`)
        export const down: Migration = ({slonik, sql}) => slonik.query(sql\`select true\`)
      `,
      down: null,
    },
    js: {
      up: dedent`
        exports.up = ({slonik, sql}) => slonik.query(sql\`select true\`)
        exports.down = ({slonik, sql}) => slonik.query(sql\`select true\`)
      `,
      down: null,
    },
  })

  const migrator: SlonikMigratorCLI = {
    up: (name?: string) => umzug.up(name).then(map(pick(['file', 'path']))),
    down: (name?: string) => umzug.down(name).then(map(pick(['file', 'path']))),
    create: (nameAndExtensions: string) => {
      const explicitExtension = supportedExtensions.find(ex => extname(nameAndExtensions) === `.${ex}`)
      const name = explicitExtension
        ? nameAndExtensions.slice(0, nameAndExtensions.lastIndexOf(`.${explicitExtension}`))
        : nameAndExtensions

      const extension =
        explicitExtension ||
        readdirSync(migrationsPath)
          .reverse()
          .map(filename => supportedExtensions.find(ex => extname(filename) === `.${ex}`))
          .find(Boolean) ||
        'sql'

      const timestamp = new Date()
        .toISOString()
        .replace(/\W/g, '-')
        .replace(/-\d\d-\d\d\dZ/, '')

      const filename = `${timestamp}.${name}.${extension}`
      const downDir = join(migrationsPath, 'down')
      const {up, down} = templates(name)[extension as SupportedExtension]

      const upPath = join(migrationsPath, filename)
      writeFileSync(upPath, up + EOL, 'utf8')

      if (down) {
        mkdirSync(downDir, {recursive: true})
        writeFileSync(join(downDir, filename), down + EOL, 'utf8')
      }

      return upPath
    },
  }
  /* istanbul ignore if */
  if (require.main === mainModule) {
    const [command, name] = process.argv.slice(2)
    if (command in migrator) {
      migrator[command as keyof SlonikMigratorCLI](name)
    } else {
      const info = {'commands available': Object.keys(migrator), 'command from cli args': command}
      throw `command not found. ${inspect(info, {breakLength: Infinity})}`
    }
  }

  return migrator
}
