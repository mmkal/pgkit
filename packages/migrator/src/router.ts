import {sql} from '@pgkit/client'
import * as trpcServer from '@trpc/server'
import {readFile} from 'fs/promises'
import z from 'zod'
import {Migrator} from './migrator'
import {Confirm} from './types'

export interface MigratorRouterContext {
  migrator: Migrator
  confirm: Confirm
}

/** Based on `cleye`'s `HelpOptions` */
export interface MigratorRouterMeta {
  /** Description of the script or command to display in `--help` output. */
  description?: string
  /** Usage code examples to display in `--help` output. */
  usage?: false | string | string[]
  /** Example code snippets to display in `--help` output. */
  examples?: string | string[]
}

/** Helper type that IMO should exist in trpc. Basically a type which a trpc-procedure with context of type `Ctx` will satisfy */
export type TRPCProcedureLike<Ctx> = {
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  mutation: (resolver: (opts: {ctx: Ctx; input: any}) => any) => any
}

/** Parameters needed for a helper function returning a router, which can be used as a sub-router for another. */
export interface TRPCHelperParams<Ctx> {
  /** The `trpc.procedure` function. A middleware can be used to align context types. */
  procedure: TRPCProcedureLike<Ctx>
}

export const migratorTrpc = trpcServer.initTRPC
  .context<MigratorRouterContext>()
  .meta<MigratorRouterMeta>()
  .create() satisfies TRPCHelperParams<MigratorRouterContext> // use `satisfies` to make sure the `TRPCProcedureLike` type helper is correct

/**
 * Get a router with procedures needed to manage migrations.
 *
 * @param procedure - TRPC procedure builder for the router this will be attached to. Must have the correct context type
 *
 * @example
 * import {createMigratorRouter, Migrator} from '@pgkit/migrator'
 * import {initTRPC} from '@trpc/server'
 *
 * const t = initTRPC.context<YourAppContext>().create()
 *
 * export const yourAppRouter = t.router({
 *   yourProcedeure: t.procedure.query(async () => 'Hello, world!'),
 *   migrations: getMigrationsRouter(),
 * })
 *
 * function getMigrationsRouter() {
 *   return createMigratorRouter(
 *     t.procedure.use(async ({ctx, next}) => {
 *       return next({
 *         ctx: {
 *           migrator: new Migrator(___),
 *           confirm: async (sql: string) => {
 *             return ctx.whitelistedSql.includes(sql)
 *           },
 *         }
 *       })
 *     })
 *   )
 * }
 */
export const createMigratorRouter = (procedure: TRPCProcedureLike<MigratorRouterContext> = migratorTrpc.procedure) => {
  // Take advantage of trpc being overly lax about merging routers with different contexts: https://github.com/trpc/trpc/issues/4306
  // We have used the `TRPCProcedureLike` type to ensure that the context is correct for the procedure builder, and trpc will merge the routers without checking the context
  // This means any router with a different context type can use this helper to creater a migrations sub-router, by just defining a middleware that sets the correct context
  const trpc = {router: migratorTrpc.router, procedure: procedure} as typeof migratorTrpc

  const router = trpc.router({
    up: trpc.procedure
      .meta({description: 'Apply pending migrations'})
      .input(
        z
          .union([
            z.object({}).strict(), //
            z.object({to: z.string().describe('Only apply migrations up to this one')}),
            z.object({step: z.number().int().positive().describe('Apply this many migrations')}),
          ])
          .optional(),
      )
      .mutation(async ({input, ctx}) => {
        return ctx.migrator.up(input)
      }),
    create: trpc.procedure
      .meta({description: 'Create a new migration file'})
      .input(
        z.object({
          content: z
            .string()
            .optional()
            .describe(
              'SQL content of the migration. If not specified, content will be generated based on the calculated diff between the existing migrations and the current database state.',
            ),
          name: z
            .string()
            .optional()
            .describe(
              'Name of the migration file. If not specified, a name will be generated based on the content of the migraiton',
            ),
        }),
      )
      .mutation(async ({input, ctx}) => {
        return ctx.migrator.create(input)
      }),
    list: trpc.procedure
      .meta({
        description: 'List migrations, along with their status, file path and content',
      })
      .input(
        z.object({
          query: z
            .string()
            .optional()
            .describe('Search query - migrations with names containing this string will be returned'),
          status: z.enum(['pending', 'executed']).optional().describe('Filter by status'),
          result: z
            .enum(['first', 'last', 'one', 'maybeOne', 'all'])
            .default('all')
            .describe('Which result(s) to return'),
          output: z
            .enum(['name', 'path', 'content', 'object'])
            .default('object')
            .describe('Result properties to return'),
        }),
      )
      .query(async ({input, ctx}) => {
        const list = await ctx.migrator.list()
        const results = list
          .filter(m => {
            return m.name.includes(input.query || '') && m.status === (input.status || m.status)
          })
          .map(m => {
            if (input.output === 'name') return m.name
            if (input.output === 'path') return m.path
            if (input.output === 'content') return m.content
            return m
          })

        if (input.result === 'all') return results

        if (input.result === 'one' && results.length !== 1) {
          throw new Error(
            `Expected exactly one migration matching query ${JSON.stringify(input.query)}, found ${results.length}`,
            {
              cause: {results},
            },
          )
        }

        if (input.result === 'maybeOne' && results.length !== 1) return undefined
        if (input.result === 'first') return results[0]
        if (input.result === 'last') return results.at(-1)
      }),
    latest: trpc.procedure
      .meta({
        description: 'Get the latest migration',
      })
      .input(
        z.object({
          skipCheck: z.boolean().optional().describe('Skip checking that migrations are in a valid state'),
        }),
      )
      .query(async ({input, ctx}) => {
        return ctx.migrator.latest(input)
      }),
    check: trpc.procedure
      .meta({description: 'Verify that your database is in an expected state, matching your migrations'})
      .mutation(async ({ctx}) => {
        return ctx.migrator.check()
      }),
    repair: trpc.procedure
      .meta({
        description:
          'If your migrations are not in a valid state, this will calculate the diff required to move your database to a valid state, and apply it',
      })
      .mutation(async ({ctx}) => {
        return ctx.migrator.repair({confirm: ctx.confirm})
      }),
    goto: trpc.procedure
      .meta({
        description:
          'Go "back" to a specific migration. This will calculate the diff required to get to the target migration, then apply it',
      })
      .input(
        z.object({
          name: z.string().describe('Name of the migration to go to. Use "list" to see available migrations.'),
        }),
      )
      .mutation(async ({input, ctx}) => {
        return ctx.migrator.goto({name: input.name, confirm: ctx.confirm})
      }),
    baseline: trpc.procedure
      .meta({
        description:
          'Baseline the database at the specified migration. This forcibly edits the migrations table to mark all migrations up to this point as executed. Useful for introducing the migrator to an existing database.',
      })
      .input(
        z.object({
          to: z.string().describe('Name of the migration to baseline to. Use `list` to see available migrations.'),
          purgeDisk: z.boolean().default(false).describe('Delete files subsequent to the specified migration'),
        }),
      )
      .mutation(async ({input, ctx}) => {
        return ctx.migrator.baseline({...input, to: input.to, confirm: ctx.confirm})
      }),
    rebase: trpc.procedure
      .meta({
        description:
          'Rebase the migrations from the specified migration. This deletes all migration files after this point, and replaces them with a squashed migration based on the calculated diff required to reach the current database state.',
      })
      .input(
        z.object({
          from: z
            .string()
            .describe(
              'Name of the migration to rebase from. This migration will remain, all subsequent ones will be replaced with a squashed migration. Use `list` to see available migrations.',
            ),
        }),
      )
      .mutation(async ({input, ctx}) => {
        return ctx.migrator.rebase({...input, confirm: ctx.confirm})
      }),
    definitions: trpc.router({
      filepath: trpc.procedure
        .meta({
          description: 'Get the path to the definitions file',
        })
        .query(async ({ctx}) => {
          return {
            path: ctx.migrator.definitionsFile,
            content: await readFile(ctx.migrator.definitionsFile, 'utf8').catch(() => null),
          }
        }),
      updateDb: trpc.procedure
        .meta({description: 'Update the database from the definitions file'})
        .mutation(async ({ctx}) => {
          return ctx.migrator.updateDbToMatchDefinitions({confirm: ctx.confirm})
        }),
      updateFile: trpc.procedure
        .meta({description: 'Update the definitions file from the database'})
        .mutation(async ({ctx}) => {
          return ctx.migrator.updateDefinitionsToMatchDb({confirm: ctx.confirm})
        }),
    }),
    unlock: trpc.procedure
      .meta({
        description:
          'Release the advisory lock for this migrator on the database. This is useful if the migrator is stuck due to a previous crash',
      })
      .mutation(async ({ctx}) => {
        return ctx.migrator.unlock({confirm: ctx.confirm})
      }),
    wipe: trpc.procedure
      .meta({
        description: 'Wipe the database - remove all tables, views etc.',
      })
      .mutation(async ({ctx}) => {
        return ctx.migrator.wipe({confirm: ctx.confirm})
      }),
    sql: trpc.procedure
      .meta({
        description:
          'Query the database. Not strictly related to migrations, but can be used for debugging. Use with caution!',
      })
      .input(
        z.tuple([
          z.string(),
          z.object({
            singlequote: z
              .string()
              .describe("Character to use in place of ' - use to avoid having to do bash quote-escaping")
              .optional(),
            doublequote: z
              .string()
              .describe('Character to use in place of " - use to avoid having to do bash quote-escaping')
              .optional(),
            method: z
              .enum(['any', 'many', 'one', 'maybeOne', 'query', 'anyFirst', 'oneFirst', 'maybeOneFirst'])
              .default('any'),
          }),
        ]),
      )
      .mutation(async ({input: [query, options], ctx}) => {
        if (options.singlequote) query = query.replaceAll(options.singlequote, `'`)
        if (options.doublequote) query = query.replaceAll(options.doublequote, `"`)

        return ctx.migrator.client[options.method](sql.raw(query))
      }),
  })

  return router
}
