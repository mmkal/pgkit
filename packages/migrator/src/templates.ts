export const sqlUp = `raise 'up migration not implemented'
`

export const sqlDown = `raise 'down migration not implemented'
`

export const typescript = `
import {Migration} from '@slonik/migrator'

export const up: Migration = async ({context: {connection, sql}}) => {
  await connection.query(sql\`raise 'up migration not implemented'\`)
}

export const down: Migration = async ({context: {connection, sql}}) => {
  await connection.query(sql\`raise 'down migration not implemented'\`)
}
`.trimLeft()

export const javascript = `
/** @type {import('@sloink/migrator').Migration} */
exports.up = async ({context: {connection, sql}}) => {
  await connection.query(sql\`raise 'up migration not implemented'\`)
}

/** @type {import('@sloink/migrator').Migration} */
exports.down = async ({context: {connection, sql}}) => {
  await connection.query(sql\`raise 'down migration not implemented'\`)
}
`.trimLeft()
