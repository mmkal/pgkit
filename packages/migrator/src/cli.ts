import * as prompt from '@inquirer/prompts'
import * as childProcess from 'child_process'
import * as fs from 'fs'
import * as os from 'os'
import * as path from 'path'
import * as colors from 'picocolors'
import {createCli} from 'trpc-cli'
import {Migrator} from './migrator'
import {createMigratorRouter} from './router'

export const getMigratorFromEnv = () => {
  return new Migrator({
    client: process.env.PGKIT_CONNECTION_STRING || 'postgresql://postgres:postgres@localhost:5432/postgres',
    migrationsPath: process.env.PGKIT_MIGRATIONS_PATH || path.join(process.cwd(), 'migrations'),
    migrationTableName: process.env.PGKIT_MIGRATIONS_TABLE_NAME || 'pgkit_migrations',
  })
}

export const createMigratorCli = (migrator: Migrator) => {
  return createCli({
    router: createMigratorRouter(),
    context: {migrator, confirm},
  })
}

export type MigratorCLI = ReturnType<typeof createMigratorCli>

export const confirm = async (input: string, options?: {readonly?: boolean}): Promise<string | null> => {
  let currentInput = input

  while (currentInput.trim()) {
    const message = `${colors.underline('Please confirm you want to run the following')}:\n\n${currentInput}`
    const editors = [
      ...new Set([process.env.VISUAL, process.env.EDITOR, 'cursor', 'code', 'vi', 'emacs', 'nano', 'notepad']),
    ].filter(name => {
      return !name?.match(/\W/) && childProcess.execSync(`sh -c 'which ${name} || echo ""'`).toString().trim()
    })
    const choice = await prompt.select({
      message,
      choices: [
        {name: 'Yes', value: 'yes'},
        {name: 'No', value: 'no'},
        ...editors.map(name => ({
          name: `Edit with ${name}`,
          disabled: options?.readonly,
          description: `Save and close the file after editing. You'll be asked again if you want to run the edit migration afterwards.`,
          value: `edit-${name}`,
        })),
      ],
    })

    if (choice === 'yes') return currentInput
    if (choice === 'no') return null

    currentInput = await editTempFile(currentInput, choice.replace('edit-', ''))
  }

  return null
}

async function editTempFile(input: string, editor = process.env.EDITOR || 'vi'): Promise<string> {
  const tempFile = path.join(os.tmpdir(), `pgkit-migrator`, `changes-${Date.now()}.sql`)
  fs.mkdirSync(path.dirname(tempFile), {recursive: true})
  fs.writeFileSync(tempFile, input.trim() + '\n')

  try {
    await new Promise<void>((resolve, reject) => {
      const child = childProcess.spawn(editor, [tempFile], {stdio: 'inherit'})
      child.on('exit', code => {
        if (code === 0) resolve()
        else reject(new Error(`Editor exited with code ${code}`))
      })
    })

    // Wait for user input to ensure the file has been saved
    await prompt.input({message: 'Press Enter when you have finished editing'})

    const editedContent = fs.readFileSync(tempFile, 'utf8')
    return editedContent.trim()
  } finally {
    fs.unlinkSync(tempFile)
  }
}
