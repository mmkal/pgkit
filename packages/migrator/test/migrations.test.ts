import {join} from 'path'
import {range} from 'lodash'
import {createPool, sql} from 'slonik'
import {setupSlonikMigrator} from '../src'
import * as dedent from 'dedent'
import {fsSyncer} from 'fs-syncer'

const slonik = createPool('postgresql://postgres:postgres@localhost:5433/postgres', {idleTimeout: 1})

const millisPerDay = 1000 * 60 * 60 * 24
const fakeDates = range(0, 100).map(days => new Date(new Date('2000').getTime() + days * millisPerDay).toISOString())

const toISOSpy = jest.spyOn(Date.prototype, 'toISOString')
toISOSpy.mockImplementation(() => fakeDates[toISOSpy.mock.calls.length - 1])

describe('run migrations', () => {
  const migrationsPath = join(__dirname, 'generated/run/migrations')

  const mockLogger = jest.fn()
  const log = (...args: any[]) => {
    if (JSON.stringify([args]) === JSON.stringify(mockLogger.mock.calls.slice(-1))) {
      return
    }
    mockLogger(...args)
  }

  const migrator = setupSlonikMigrator({
    slonik,
    migrationsPath,
    migrationTableName: 'migration_test',
    log,
  })

  const syncer = fsSyncer(migrationsPath, {
    '01.one.sql': 'create table migration_test_1(id int)',
    '02.two.sql': 'create table migration_test_2(id int)',
    '03.three.js': dedent`
      module.exports.up = ({slonik, sql}) => slonik.query(sql\`create table migration_test_3(id int)\`)
      module.exports.down = ({slonik, sql}) => slonik.query(sql\`drop table migration_test_3\`)
    `,
    '04.four.ts': dedent`
      import {Migration} from '../..'

      export const up: Migration = ({slonik, sql}) => slonik.query(sql\`create table migration_test_4(id int)\`)
      export const down: Migration = ({slonik, sql}) => slonik.query(sql\`drop table migration_test_4\`)
    `,
    down: {
      '01.one.sql': 'drop table migration_test_1',
      '02.two.sql': 'drop table migration_test_2',
    },
  })

  const deleteTables = () =>
    slonik.query(sql`
      drop table if exists migration_test;
      drop table if exists migration_test_;
      drop table if exists migration_test_1;
      drop table if exists migration_test_2;
      drop table if exists migration_test_3;
      drop table if exists migration_test_4;
    `)

  beforeAll(async () => {
    await deleteTables()
    await syncer.sync()
  })

  const migrationTables = () =>
    slonik.anyFirst(sql`
      select tablename
      from pg_catalog.pg_tables
      where tablename like 'migration_test_%'
      order by tablename
    `)

  test('up and down', async () => {
    expect(await migrationTables()).toEqual([])

    const executed = () => migrator.executed().then(list => list.map(x => x.file))
    const pending = () => migrator.pending().then(list => list.map(x => x.file))

    log(`getting pending migrations before initial 'up'`)

    const allMigrations = await pending()
    expect(allMigrations).toMatchInlineSnapshot(`
      Array [
        "01.one.sql",
        "02.two.sql",
        "03.three.js",
        "04.four.ts",
      ]
    `)

    await migrator.up()
    const allTables = await migrationTables()
    expect(allTables).toMatchInlineSnapshot(`
      Array [
        "migration_test_1",
        "migration_test_2",
        "migration_test_3",
        "migration_test_4",
      ]
    `)

    expect(await executed()).toEqual(allMigrations)
    expect(await pending()).toEqual([])

    // make sure `.down` with no arg reverts only the last migration
    await migrator.down()
    expect(await migrationTables()).toEqual(allTables.slice(0, -1))
    expect(await executed()).toEqual(allMigrations.slice(0, -1))
    expect(await pending()).toEqual(allMigrations.slice(-1))

    await migrator.down()
    expect(await migrationTables()).toEqual(allTables.slice(0, -2))
    expect(await executed()).toEqual(allMigrations.slice(0, -2))
    expect(await pending()).toEqual(allMigrations.slice(-2))

    await migrator.up()
    expect(await migrationTables()).toEqual(allTables)
    expect(await executed()).toEqual(allMigrations)
    expect(await pending()).toEqual([])

    await migrator.down('02.two.sql')
    expect(await executed()).toMatchInlineSnapshot(`
      Array [
        "01.one.sql",
      ]
    `)

    await migrator.down('0')
    expect(await executed()).toEqual([])

    await migrator.up('03.three.js')
    expect(await executed()).toMatchInlineSnapshot(`
      Array [
        "01.one.sql",
        "02.two.sql",
        "03.three.js",
      ]
    `)

    expect(
      mockLogger.mock.calls.map(msg => {
        const json = JSON.stringify(msg)
        return JSON.parse(json.replace(/\d\.\d\d\ds/g, '?.???s'))
      }),
    ).toMatchSnapshot()
  })
})

// https://github.com/gajus/slonik/issues/63#issuecomment-500889445
afterAll(() => new Promise(r => setTimeout(r, 1)))
