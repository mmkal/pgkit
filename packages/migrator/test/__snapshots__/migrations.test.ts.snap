// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`run migrations up and down 5`] = `
Array [
  Array [
    "getting pending migrations before initial 'up'",
  ],
  Array [
    "migrations in database:",
    Array [],
  ],
  Array [
    "== 01.one: migrating =======",
  ],
  Array [
    "== 01.one: migrated (?.???s)
",
  ],
  Array [
    "migrations in database:",
    Array [
      Object {
        "hash": "68a0494793",
        "name": "01.one.sql",
      },
    ],
  ],
  Array [
    "== 02.two: migrating =======",
  ],
  Array [
    "== 02.two: migrated (?.???s)
",
  ],
  Array [
    "migrations in database:",
    Array [
      Object {
        "hash": "68a0494793",
        "name": "01.one.sql",
      },
      Object {
        "hash": "a65a32e9d3",
        "name": "02.two.sql",
      },
    ],
  ],
  Array [
    "== 03.three: migrating =======",
  ],
  Array [
    "== 03.three: migrated (?.???s)
",
  ],
  Array [
    "migrations in database:",
    Array [
      Object {
        "hash": "68a0494793",
        "name": "01.one.sql",
      },
      Object {
        "hash": "a65a32e9d3",
        "name": "02.two.sql",
      },
      Object {
        "hash": "a97f6162ee",
        "name": "03.three.js",
      },
    ],
  ],
  Array [
    "== 04.four: migrating =======",
  ],
  Array [
    "== 04.four: migrated (?.???s)
",
  ],
  Array [
    "migrations in database:",
    Array [
      Object {
        "hash": "68a0494793",
        "name": "01.one.sql",
      },
      Object {
        "hash": "a65a32e9d3",
        "name": "02.two.sql",
      },
      Object {
        "hash": "a97f6162ee",
        "name": "03.three.js",
      },
      Object {
        "hash": "a6f0526ce6",
        "name": "04.four.ts",
      },
    ],
  ],
  Array [
    "== 04.four: reverting =======",
  ],
  Array [
    "== 04.four: reverted (?.???s)
",
  ],
  Array [
    "migrations in database:",
    Array [
      Object {
        "hash": "68a0494793",
        "name": "01.one.sql",
      },
      Object {
        "hash": "a65a32e9d3",
        "name": "02.two.sql",
      },
      Object {
        "hash": "a97f6162ee",
        "name": "03.three.js",
      },
    ],
  ],
  Array [
    "== 03.three: reverting =======",
  ],
  Array [
    "== 03.three: reverted (?.???s)
",
  ],
  Array [
    "migrations in database:",
    Array [
      Object {
        "hash": "68a0494793",
        "name": "01.one.sql",
      },
      Object {
        "hash": "a65a32e9d3",
        "name": "02.two.sql",
      },
    ],
  ],
  Array [
    "== 03.three: migrating =======",
  ],
  Array [
    "== 03.three: migrated (?.???s)
",
  ],
  Array [
    "migrations in database:",
    Array [
      Object {
        "hash": "68a0494793",
        "name": "01.one.sql",
      },
      Object {
        "hash": "a65a32e9d3",
        "name": "02.two.sql",
      },
      Object {
        "hash": "a97f6162ee",
        "name": "03.three.js",
      },
    ],
  ],
  Array [
    "== 04.four: migrating =======",
  ],
  Array [
    "== 04.four: migrated (?.???s)
",
  ],
  Array [
    "migrations in database:",
    Array [
      Object {
        "hash": "68a0494793",
        "name": "01.one.sql",
      },
      Object {
        "hash": "a65a32e9d3",
        "name": "02.two.sql",
      },
      Object {
        "hash": "a97f6162ee",
        "name": "03.three.js",
      },
      Object {
        "hash": "a6f0526ce6",
        "name": "04.four.ts",
      },
    ],
  ],
  Array [
    "== 04.four: reverting =======",
  ],
  Array [
    "== 04.four: reverted (?.???s)
",
  ],
  Array [
    "migrations in database:",
    Array [
      Object {
        "hash": "68a0494793",
        "name": "01.one.sql",
      },
      Object {
        "hash": "a65a32e9d3",
        "name": "02.two.sql",
      },
      Object {
        "hash": "a97f6162ee",
        "name": "03.three.js",
      },
    ],
  ],
  Array [
    "== 03.three: reverting =======",
  ],
  Array [
    "== 03.three: reverted (?.???s)
",
  ],
  Array [
    "migrations in database:",
    Array [
      Object {
        "hash": "68a0494793",
        "name": "01.one.sql",
      },
      Object {
        "hash": "a65a32e9d3",
        "name": "02.two.sql",
      },
    ],
  ],
  Array [
    "== 02.two: reverting =======",
  ],
  Array [
    "== 02.two: reverted (?.???s)
",
  ],
  Array [
    "migrations in database:",
    Array [
      Object {
        "hash": "68a0494793",
        "name": "01.one.sql",
      },
    ],
  ],
  Array [
    "== 01.one: reverting =======",
  ],
  Array [
    "== 01.one: reverted (?.???s)
",
  ],
  Array [
    "migrations in database:",
    Array [],
  ],
  Array [
    "== 01.one: migrating =======",
  ],
  Array [
    "== 01.one: migrated (?.???s)
",
  ],
  Array [
    "migrations in database:",
    Array [
      Object {
        "hash": "68a0494793",
        "name": "01.one.sql",
      },
    ],
  ],
  Array [
    "== 02.two: migrating =======",
  ],
  Array [
    "== 02.two: migrated (?.???s)
",
  ],
  Array [
    "migrations in database:",
    Array [
      Object {
        "hash": "68a0494793",
        "name": "01.one.sql",
      },
      Object {
        "hash": "a65a32e9d3",
        "name": "02.two.sql",
      },
    ],
  ],
  Array [
    "== 03.three: migrating =======",
  ],
  Array [
    "== 03.three: migrated (?.???s)
",
  ],
  Array [
    "migrations in database:",
    Array [
      Object {
        "hash": "68a0494793",
        "name": "01.one.sql",
      },
      Object {
        "hash": "a65a32e9d3",
        "name": "02.two.sql",
      },
      Object {
        "hash": "a97f6162ee",
        "name": "03.three.js",
      },
    ],
  ],
]
`;
